clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -118,6 +130,24 @@ public void assign(List<TopicPartition> partitions) {
         this.assignment.keySet().retainAll(this.userAssignment);
     }
 
+    public void subscribe(Pattern pattern) {
+        if (!this.subscription.isEmpty() || !this.userAssignment.isEmpty())
```

Comment:
For some reason I think explicitly unsubscribing topic before subscribing via pattern is better, leaves less room for user error. If you think it is still OK, I can make the change. Let me know.